File: cgse-lab1-java/src/main/java/com/arula/lab1/service/UserService.java
/**
 * CGSE-A Lab 1 - Java/Spring Boot Starter Code
 * UserService - Contains INTENTIONAL security vulnerabilities
 *
 * WARNING: DO NOT use in production!
 *
 * VULNERABILITIES TO FIND:
 * 1. Line ~55: SQL Injection via JPQL string concatenation
 * 2. Line ~65: SQL Injection via LIKE clause
 * 3. Line ~75: PII/Password logging
 * 4. Line ~85: Mass assignment via BeanUtils.copyProperties
 * 5. Line ~95: Timing attack in credential verification
 * 6. Line ~100: Plain text password comparison
 */

package com.arula.lab1.service;

import com.arula.lab1.model.User;
import com.arula.lab1.repository.UserRepository;
import jakarta.persistence.EntityManager;
import jakarta.persistence.PersistenceContext;
import org.springframework.beans.BeanUtils;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class UserService {

    private final UserRepository userRepository;

    @PersistenceContext
    private EntityManager entityManager;

    public UserService(UserRepository userRepository) {
        this.userRepository = userRepository;
    }

    public List<User> findAll() {
        return userRepository.findAll();
    }

    public User findById(Long id) {
        return userRepository.findById(id).orElse(null);
    }

    // VULNERABILITY 1: SQL Injection via JPQL concatenation
    public User findByEmailUnsafe(String email) {
        String jpql = "SELECT u FROM User u WHERE u.email = '" + email + "'";
        try {
            return entityManager.createQuery(jpql, User.class).getSingleResult();
        } catch (Exception e) {
            System.out.println("Error: " + e.getMessage()); // Info disclosure
            return null;
        }
    }

    // VULNERABILITY 2: SQL Injection via LIKE clause
    public List<User> searchByName(String namePattern) {
        String jpql = "SELECT u FROM User u WHERE u.name LIKE '%" + namePattern + "%'";
        return entityManager.createQuery(jpql, User.class).getResultList();
    }

    // VULNERABILITY 3: PII and password logging
    public User createUser(User user) {
        System.out.println("Creating user: " + user.toString());
        System.out.println("Password: " + user.getPassword()); // CRITICAL!
        return userRepository.save(user);
    }

    // VULNERABILITY 4: Mass assignment
    public User updateUser(Long id, User updatedUser) {
        User user = findById(id);
        if (user == null) {
            throw new RuntimeException("User not found");
        }
        // Copies ALL fields including role, isAdmin, etc.
        BeanUtils.copyProperties(updatedUser, user, "id");
        return userRepository.save(user);
    }

    public void deleteUser(Long id) {
        userRepository.deleteById(id);
    }

    // VULNERABILITY 5: Timing attack
    // VULNERABILITY 6: Plain text password comparison
    public boolean verifyCredentials(String email, String password) {
        User user = userRepository.findByEmail(email).orElse(null);

        // Timing attack: fast return if user doesn't exist
        if (user == null) {
            return false;
        }

        // Plain text comparison (no hashing)
        return user.getPassword().equals(password);
    }

    // VULNERABILITY: User enumeration
    public boolean emailExists(String email) {
        return userRepository.existsByEmail(email);
    }
}

File: cgse-lab1-angular/src/app/services/user.service.ts
/**
 * CGSE-A Lab 1 - Angular Starter Code
 * UserService - Contains INTENTIONAL security vulnerabilities
 *
 * WARNING: DO NOT use in production!
 *
 * VULNERABILITIES TO FIND:
 * 1. Line ~50: URL parameter injection (string interpolation)
 * 2. Line ~55: PII logging to console
 * 3. Line ~60: Mass assignment (accepts any fields)
 * 4. Line ~75: User enumeration endpoint
 * 5. Line ~65: No error handling
 */

import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { User } from '../models/user.model';

@Injectable({
  providedIn: 'root'
})
export class UserService {
  private apiUrl = 'http://localhost:3000/api';

  constructor(private http: HttpClient) {}

  getUsers(): Observable<User[]> {
    return this.http.get<User[]>(`${this.apiUrl}/users`);
  }

  getUserById(id: string): Observable<User> {
    return this.http.get<User>(`${this.apiUrl}/users/${id}`);
  }

  // VULNERABILITY 1: SQL-like injection via URL parameter
  // Direct string interpolation allows injection attacks
  getUserByFilter(filter: string): Observable<User[]> {
    // VULNERABLE: Direct interpolation without encoding
    return this.http.get<User[]>(`${this.apiUrl}/users?filter=${filter}`);
  }

  // VULNERABILITY 2: PII logging
  searchUsers(query: string): Observable<User[]> {
    console.log('Search query:', query); // May log PII
    return this.http.get<User[]>(`${this.apiUrl}/users/search?q=${query}`);
  }

  // VULNERABILITY 3: No input validation, PII logged
  createUser(user: User): Observable<User> {
    console.log('Creating user:', user); // Logs entire user object including password
    return this.http.post<User>(`${this.apiUrl}/users`, user);
  }

  // VULNERABILITY 4: Mass assignment - accepts any fields
  updateUser(id: string, user: Partial<User>): Observable<User> {
    // Could update role, isAdmin, etc.
    return this.http.put<User>(`${this.apiUrl}/users/${id}`, user);
  }

  deleteUser(id: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/users/${id}`);
  }

  // VULNERABILITY 5: Information disclosure - user enumeration
  checkEmailExists(email: string): Observable<boolean> {
    // Allows attackers to enumerate valid emails
    return this.http.get<boolean>(`${this.apiUrl}/users/check-email?email=${email}`);
  }
}
